File: android-native/app/src/main/java/com/example/kpkn/telemetry/TelemetryHelper.kt
package com.example.kpkn.telemetry

import android.content.Context
import android.os.SystemClock
import com.example.kpkn.BuildConfig
import kotlin.math.roundToInt

/**
 * Helper class for logging telemetry events with common parameters
 */
class TelemetryHelper(private val context: Context) {
    
    private val telemetry = KpknTelemetry.getInstance(context)
    
    /**
     * Log app lifecycle events
     */
    fun logAppOpen() {
        context.logKpknEvent(
            TelemetryEvents.APP_OPEN,
            TelemetryParameters.SCREEN_NAME to "main",
            "app_version" to BuildConfig.VERSION_NAME,
            "app_version_code" to BuildConfig.VERSION_CODE
        )
    }
    
    fun logAppForeground() {
        context.logKpknEvent(
            TelemetryEvents.APP_FOREGROUND,
            TelemetryParameters.SCREEN_NAME to "foreground"
        )
    }
    
    fun logAppBackground() {
        context.logKpknEvent(
            TelemetryEvents.APP_BACKGROUND,
            TelemetryParameters.SCREEN_NAME to "background"
        )
    }
    
    /**
     * Log authentication events
     */
    fun logLoginSuccess(userId: String? = null) {
        context.logKpknEvent(
            TelemetryEvents.LOGIN_SUCCESS,
            TelemetryParameters.USER_ID to userId,
            TelemetryParameters.SUCCESS to true
        )
    }
    
    fun logLoginFailed(error: String? = null) {
        context.logKpknEvent(
            TelemetryEvents.LOGIN_FAILED,
            TelemetryParameters.ERROR_MESSAGE to error,
            TelemetryParameters.SUCCESS to false
        )
    }
    
    fun logLogout() {
        context.logKpknEvent(
            TelemetryEvents.LOGOUT,
            TelemetryParameters.SUCCESS to true
        )
    }
    
    /**
     * Log workout events with timing
     */
    fun logWorkoutStart(workoutId: String, workoutName: String, workoutType: String) {
        context.logKpknEvent(
            TelemetryEvents.WORKOUT_START,
            TelemetryParameters.WORKOUT_ID to workoutId,
            TelemetryParameters.WORKOUT_NAME to workoutName,
            TelemetryParameters.WORKOUT_TYPE to workoutType,
            TelemetryParameters.SUCCESS to true
        )
    }
    
    fun logWorkoutEnd(
        workoutId: String,
        durationMs: Long,
        exercisesCount: Int,
        setsCount: Int
    ) {
        val durationSeconds = durationMs / 1000
        context.logKpknEvent(
            TelemetryEvents.WORKOUT_END,
            TelemetryParameters.WORKOUT_ID to workoutId,
            TelemetryParameters.DURATION to durationSeconds,
            TelemetryParameters.ACTION to "complete",
            TelemetryParameters.SUCCESS to true,
            "exercises_count" to exercisesCount,
            "sets_count" to setsCount
        )
    }
    
    fun logWorkoutPause(workoutId: String) {
        context.logKpknEvent(
            TelemetryEvents.WORKOUT_PAUSE,
            TelemetryParameters.WORKOUT_ID to workoutId,
            TelemetryParameters.ACTION to "pause"
        )
    }
    
    fun logWorkoutResume(workoutId: String) {
        context.logKpknEvent(
            TelemetryEvents.WORKOUT_RESUME,
            TelemetryParameters.WORKOUT_ID to workoutId,
            TelemetryParameters.ACTION to "resume"
        )
    }
    
    fun logSetComplete(
        exerciseId: String,
        exerciseName: String,
        setNumber: Int,
        reps: Int,
        weight: Double?
    ) {
        context.logKpknEvent(
            TelemetryEvents.SET_COMPLETE,
            TelemetryParameters.EXERCISE_ID to exerciseId,
            TelemetryParameters.EXERCISE_NAME to exerciseName,
            TelemetryParameters.SET_NUMBER to setNumber,
            TelemetryParameters.REPS to reps,
            TelemetryParameters.WEIGHT to weight
        )
    }
    
    fun logRestStart(workoutId: String, restTimeSeconds: Int) {
        context.logKpknEvent(
            TelemetryEvents.REST_START,
            TelemetryParameters.WORKOUT_ID to workoutId,
            TelemetryParameters.REST_TIME to restTimeSeconds
        )
    }
    
    fun logRestEnd(workoutId: String, actualRestTimeSeconds: Int) {
        context.logKpknEvent(
            TelemetryEvents.REST_END,
            TelemetryParameters.WORKOUT_ID to workoutId,
            TelemetryParameters.REST_TIME to actualRestTimeSeconds,
            TelemetryParameters.ACTION to "complete"
        )
    }
    
    /**
     * Log nutrition events
     */
    fun logNutritionOpen() {
        context.logKpknEvent(
            TelemetryEvents.NUTRITION_OPEN,
            TelemetryParameters.SCREEN_NAME to "nutrition"
        )
    }
    
    fun logMealLogStart(mealType: String) {
        context.logKpknEvent(
            TelemetryEvents.MEAL_LOG_START,
            TelemetryParameters.MEAL_TYPE to mealType
        )
    }
    
    fun logMealLogComplete(
        mealId: String,
        mealType: String,
        totalCalories: Double,
        protein: Double,
        carbs: Double,
        fat: Double
    ) {
        context.logKpknEvent(
            TelemetryEvents.MEAL_LOG_COMPLETE,
            TelemetryParameters.MEAL_ID to mealId,
            TelemetryParameters.MEAL_TYPE to mealType,
            TelemetryParameters.CALORIES to totalCalories.roundToInt(),
            TelemetryParameters.PROTEIN to protein.roundToInt(),
            TelemetryParameters.CARBS to carbs.roundToInt(),
            TelemetryParameters.FAT to fat.roundToInt(),
            TelemetryParameters.SUCCESS to true
        )
    }
    
    fun logFoodSearch(query: String) {
        context.logKpknEvent(
            TelemetryEvents.FOOD_SEARCH,
            TelemetryParameters.ACTION to "search",
            TelemetryParameters.AI_INPUT to query
        )
    }
    
    fun logFoodItemAdd(foodId: String, foodName: String, calories: Double?) {
        context.logKpknEvent(
            TelemetryEvents.FOOD_ITEM_ADD,
            TelemetryParameters.FOOD_ID to foodId,
            TelemetryParameters.FOOD_NAME to foodName,
            TelemetryParameters.CALORIES to calories?.roundToInt()
        )
    }
    
    fun logNutritionPhotoParse() {
        context.logKpknEvent(
            TelemetryEvents.NUTRITION_PARSE_PHOTO,
            TelemetryParameters.NUTRITION_METHOD to "photo"
        )
    }
    
    fun logNutritionParseSuccess(foodItemsCount: Int) {
        context.logKpknEvent(
            TelemetryEvents.NUTRITION_PARSE_SUCCESS,
            TelemetryParameters.ACTION to "success",
            TelemetryParameters.SUCCESS to true,
            "food_items_count" to foodItemsCount
        )
    }
    
    fun logNutritionParseFailed(error: String? = null) {
        context.logKpknEvent(
            TelemetryEvents.NUTRITION_PARSE_FAILED,
            TelemetryParameters.ERROR_MESSAGE to error,
            TelemetryParameters.SUCCESS to false
        )
    }
    
    /**
     * Log program events
     */
    fun logProgramSelect(programId: String, programName: String, programType: String) {
        context.logKpknEvent(
            TelemetryEvents.PROGRAM_SELECT,
            TelemetryParameters.PROGRAM_ID to programId,
            TelemetryParameters.PROGRAM_NAME to programName,
            TelemetryParameters.PROGRAM_TYPE to programType
        )
    }
    
    fun logProgramStart(programId: String, week: Int, day: Int) {
        context.logKpknEvent(
            TelemetryEvents.PROGRAM_START,
            TelemetryParameters.PROGRAM_ID to programId,
            TelemetryParameters.PROGRAM_WEEK to week,
            TelemetryParameters.PROGRAM_DAY to day,
            TelemetryParameters.ACTION to "start"
        )
    }
    
    fun logProgramComplete(
        programId: String,
        durationMs: Long,
        totalWorkouts: Int,
        completedWorkouts: Int
    ) {
        val durationDays = durationMs / (1000 * 60 * 60 * 24)
        context.logKpknEvent(
            TelemetryEvents.PROGRAM_COMPLETE,
            TelemetryParameters.PROGRAM_ID to programId,
            TelemetryParameters.DURATION to durationDays,
            TelemetryParameters.ACTION to "complete",
            "total_workouts" to totalWorkouts,
            "completed_workouts" to completedWorkouts,
            TelemetryParameters.SUCCESS to true
        )
    }
    
    /**
     * Log exercise events
     */
    fun logExerciseSearch(query: String) {
        context.logKpknEvent(
            TelemetryEvents.EXERCICE_SEARCH,
            TelemetryParameters.ACTION to "search",
            TelemetryParameters.AI_INPUT to query
        )
    }
    
    fun logExerciseViewDetails(exerciseId: String, exerciseName: String, category: String?) {
        context.logKpknEvent(
            TelemetryEvents.EXERCICE_VIEW_DETAILS,
            TelemetryParameters.EXERCISE_ID to exerciseId,
            TelemetryParameters.EXERCISE_NAME to exerciseName,
            TelemetryParameters.EXERCISE_CATEGORY to category
        )
    }
    
    fun logExerciseAddToWorkout(exerciseId: String, exerciseName: String) {
        context.logKpknEvent(
            TelemetryEvents.EXERCICE_ADD_TO_WORKOUT,
            TelemetryParameters.EXERCISE_ID to exerciseId,
            TelemetryParameters.EXERCISE_NAME to exerciseName
        )
    }
    
    /**
     * Log navigation events
     */
    fun logNavigation(from: String, to: String) {
        context.logKpknEvent(
            TelemetryEvents.NAVIGATION,
            TelemetryParameters.NAVIGATION_FROM to from,
            TelemetryParameters.NAVIGATION_TO to to
        )
    }
    
    fun logDeepLinkOpen(path: String) {
        context.logKpknEvent(
            TelemetryEvents.DEEP_LINK_OPEN,
            TelemetryParameters.DEEP_LINK_PATH to path
        )
    }
    
    /**
     * Log AI events
     */
    fun logAIRequest(type: String, model: String, input: String? = null) {
        context.logKpknEvent(
            TelemetryEvents.AI_REQUEST,
            TelemetryParameters.AI_TYPE to type,
            TelemetryParameters.AI_MODEL to model,
            TelemetryParameters.AI_INPUT to input
        )
    }
    
    fun logAISuccess(type: String, model: String, output: String? = null) {
        context.logKpknEvent(
            TelemetryEvents.AI_SUCCESS,
            TelemetryParameters.AI_TYPE to type,
            TelemetryParameters.AI_MODEL to model,
            TelemetryParameters.AI_OUTPUT to output?.take(100), // Limit output length
            TelemetryParameters.SUCCESS to true
        )
    }
    
    fun logAIFailed(type: String, model: String, error: String? = null) {
        context.logKpknEvent(
            TelemetryEvents.AI_FAILED,
            TelemetryParameters.AI_TYPE to type,
            TelemetryParameters.AI_MODEL to model,
            TelemetryParameters.ERROR_MESSAGE to error,
            TelemetryParameters.SUCCESS to false
        )
    }
    
    /**
     * Log performance traces
     */
    class PerformanceTrace(private val telemetry: KpknTelemetry, private val name: String) {
        private val startTime = SystemClock.elapsedRealtime()
        
        fun stop() {
            val duration = SystemClock.elapsedRealtime() - startTime
            telemetry.logEvent("performance_trace", 
                TelemetryParameters.ACTION to name,
                TelemetryParameters.DURATION to duration,
                TelemetryParameters.SUCCESS to true
            )
        }
    }
    
    fun startPerformanceTrace(name: String): PerformanceTrace {
        return PerformanceTrace(telemetry, name)
    }
    
    /**
     * Log global errors
     */
    fun logGlobalError(error: Throwable, context: String? = null) {
        telemetry.logException(error, false)
        telemetry.logEvent("global_error",
            TelemetryParameters.ERROR_MESSAGE to error.message,
            TelemetryParameters.ERROR_STACK_TRACE to error.stackTraceToString().take(500),
            TelemetryParameters.ACTION to context
        )
    }
}

File: ios-native/KPKNFit/KPKNFit/Telemetry/TelemetryHelper.swift
import Foundation
import CryptoKit

class TelemetryHelper {

    private let telemetry = KpknTelemetry.shared

    private func anonymizeText(_ text: String?) -> String {
        guard let text = text, !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return ""
        }
        let input = text.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard let data = input.data(using: .utf8) else { return "anonymized" }
        let hash = SHA256.hash(data: data)
        return hash.compactMap { String(format: "%02x", $0) }.joined().prefix(12).description
    }

    func logAppOpen() {
        logKpknEvent(
            eventName: TelemetryEvents.appOpen,
            params: [
                TelemetryParameters.screenName: "main",
                "app_version": (Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String) ?? "",
                "app_version_code": (Bundle.main.infoDictionary?["CFBundleVersion"] as? String) ?? ""
            ]
        )
    }

    func logAppForeground() {
        logKpknEvent(
            eventName: TelemetryEvents.appForeground,
            params: [
                TelemetryParameters.screenName: "foreground"
            ]
        )
    }

    func logAppBackground() {
        logKpknEvent(
            eventName: TelemetryEvents.appBackground,
            params: [
                TelemetryParameters.screenName: "background"
            ]
        )
    }

    func logLoginSuccess(userId: String? = nil) {
        logKpknEvent(
            eventName: TelemetryEvents.loginSuccess,
            params: [
                TelemetryParameters.userId: userId.map { anonymizeText($0) } as Any,
                TelemetryParameters.success: true
            ]
        )
    }

    func logLoginFailed(error: String? = nil) {
        logKpknEvent(
            eventName: TelemetryEvents.loginFailed,
            params: [
                TelemetryParameters.errorMessage: error as Any,
                TelemetryParameters.success: false
            ]
        )
    }

    func logLogout() {
        logKpknEvent(
            eventName: TelemetryEvents.logout,
            params: [
                TelemetryParameters.success: true
            ]
        )
    }

    func logWorkoutStart(workoutId: String, workoutName: String, workoutType: String) {
        logKpknEvent(
            eventName: TelemetryEvents.workoutStart,
            params: [
                TelemetryParameters.workoutId: workoutId,
                TelemetryParameters.workoutName: workoutName,
                TelemetryParameters.workoutType: workoutType,
                TelemetryParameters.success: true
            ]
        )
    }

    func logWorkoutEnd(workoutId: String, durationMs: Int, exercisesCount: Int, setsCount: Int) {
        let durationSeconds = durationMs / 1000
        logKpknEvent(
            eventName: TelemetryEvents.workoutEnd,
            params: [
                TelemetryParameters.workoutId: workoutId,
                TelemetryParameters.duration: durationSeconds,
                TelemetryParameters.action: "complete",
                TelemetryParameters.success: true,
                "exercises_count": exercisesCount,
                "sets_count": setsCount
            ]
        )
    }

    func logWorkoutPause(workoutId: String) {
        logKpknEvent(
            eventName: TelemetryEvents.workoutPause,
            params: [
                TelemetryParameters.workoutId: workoutId,
                TelemetryParameters.action: "pause"
            ]
        )
    }

    func logWorkoutResume(workoutId: String) {
        logKpknEvent(
            eventName: TelemetryEvents.workoutResume,
            params: [
                TelemetryParameters.workoutId: workoutId,
                TelemetryParameters.action: "resume"
            ]
        )
    }

    func logSetComplete(exerciseId: String, exerciseName: String, setNumber: Int, reps: Int, weight: Double?) {
        let roundedWeight = weight.map { Double(Int(($0 / 2.5).rounded())) * 2.5 }
        logKpknEvent(
            eventName: TelemetryEvents.setComplete,
            params: [
                TelemetryParameters.exerciseId: exerciseId,
                TelemetryParameters.exerciseName: exerciseName,
                TelemetryParameters.setNumber: setNumber,
                TelemetryParameters.reps: reps,
                TelemetryParameters.weight: roundedWeight as Any
            ]
        )
    }

    func logRestStart(workoutId: String, restTimeSeconds: Int) {
        logKpknEvent(
            eventName: TelemetryEvents.restStart,
            params: [
                TelemetryParameters.workoutId: workoutId,
                TelemetryParameters.restTime: restTimeSeconds
            ]
        )
    }

    func logRestEnd(workoutId: String, actualRestTimeSeconds: Int) {
        logKpknEvent(
            eventName: TelemetryEvents.restEnd,
            params: [
                TelemetryParameters.workoutId: workoutId,
                TelemetryParameters.restTime: actualRestTimeSeconds,
                TelemetryParameters.action: "complete"
            ]
        )
    }

    func logNutritionOpen() {
        logKpknEvent(
            eventName: TelemetryEvents.nutritionOpen,
            params: [
                TelemetryParameters.screenName: "nutrition"
            ]
        )
    }

    func logMealLogStart(mealType: String) {
        logKpknEvent(
            eventName: TelemetryEvents.mealLogStart,
            params: [
                TelemetryParameters.mealType: mealType
            ]
        )
    }

    func logMealLogComplete(mealId: String, mealType: String, totalCalories: Double, protein: Double, carbs: Double, fat: Double) {
        logKpknEvent(
            eventName: TelemetryEvents.mealLogComplete,
            params: [
                TelemetryParameters.mealId: mealId,
                TelemetryParameters.mealType: mealType,
                TelemetryParameters.calories: Int(totalCalories.rounded()),
                TelemetryParameters.protein: Int(protein.rounded()),
                TelemetryParameters.carbs: Int(carbs.rounded()),
                TelemetryParameters.fat: Int(fat.rounded()),
                TelemetryParameters.success: true
            ]
        )
    }

    func logFoodSearch(query: String) {
        logKpknEvent(
            eventName: TelemetryEvents.foodSearch,
            params: [
                TelemetryParameters.action: "search",
                TelemetryParameters.aiInput: anonymizeText(query)
            ]
        )
    }

    func logFoodItemAdd(foodId: String, foodName: String, calories: Double?) {
        logKpknEvent(
            eventName: TelemetryEvents.foodItemAdd,
            params: [
                TelemetryParameters.foodId: foodId,
                TelemetryParameters.foodName: anonymizeText(foodName),
                TelemetryParameters.calories: calories.map { Int($0.rounded()) } as Any
            ]
        )
    }

    func logNutritionPhotoParse() {
        logKpknEvent(
            eventName: TelemetryEvents.nutritionParsePhoto,
            params: [
                TelemetryParameters.nutritionMethod: "photo"
            ]
        )
    }

    func logNutritionParseSuccess(foodItemsCount: Int) {
        logKpknEvent(
            eventName: TelemetryEvents.nutritionParseSuccess,
            params: [
                TelemetryParameters.action: "success",
                TelemetryParameters.success: true,
                "food_items_count": foodItemsCount
            ]
        )
    }

    func logNutritionParseFailed(error: String? = nil) {
        logKpknEvent(
            eventName: TelemetryEvents.nutritionParseFailed,
            params: [
                TelemetryParameters.errorMessage: error as Any,
                TelemetryParameters.success: false
            ]
        )
    }

    func logProgramSelect(programId: String, programName: String, programType: String) {
        logKpknEvent(
            eventName: TelemetryEvents.programSelect,
            params: [
                TelemetryParameters.programId: programId,
                TelemetryParameters.programName: programName,
                TelemetryParameters.programType: programType
            ]
        )
    }

    func logProgramStart(programId: String, week: Int, day: Int) {
        logKpknEvent(
            eventName: TelemetryEvents.programStart,
            params: [
                TelemetryParameters.programId: programId,
                TelemetryParameters.programWeek: week,
                TelemetryParameters.programDay: day,
                TelemetryParameters.action: "start"
            ]
        )
    }

    func logProgramComplete(programId: String, durationMs: Int, totalWorkouts: Int, completedWorkouts: Int) {
        let durationDays = durationMs / (1000 * 60 * 60 * 24)
        logKpknEvent(
            eventName: TelemetryEvents.programComplete,
            params: [
                TelemetryParameters.programId: programId,
                TelemetryParameters.duration: durationDays,
                TelemetryParameters.action: "complete",
                "total_workouts": totalWorkouts,
                "completed_workouts": completedWorkouts,
                TelemetryParameters.success: true
            ]
        )
    }

    func logExerciseSearch(query: String) {
        logKpknEvent(
            eventName: TelemetryEvents.exerciceSearch,
            params: [
                TelemetryParameters.action: "search",
                TelemetryParameters.aiInput: anonymizeText(query)
            ]
        )
    }

    func logExerciseViewDetails(exerciseId: String, exerciseName: String, category: String?) {
        logKpknEvent(
            eventName: TelemetryEvents.exerciceViewDetails,
            params: [
                TelemetryParameters.exerciseId: exerciseId,
                TelemetryParameters.exerciseName: exerciseName,
                TelemetryParameters.exerciseCategory: category as Any
            ]
        )
    }

    func logExerciseAddToWorkout(exerciseId: String, exerciseName: String) {
        logKpknEvent(
            eventName: TelemetryEvents.exerciceAddToWorkout,
            params: [
                TelemetryParameters.exerciseId: exerciseId,
                TelemetryParameters.exerciseName: exerciseName
            ]
        )
    }

    func logNavigation(from: String, to: String) {
        logKpknEvent(
            eventName: TelemetryEvents.navigation,
            params: [
                TelemetryParameters.navigationFrom: from,
                TelemetryParameters.navigationTo: to
            ]
        )
    }

    func logDeepLinkOpen(path: String) {
        logKpknEvent(
            eventName: TelemetryEvents.deepLinkOpen,
            params: [
                TelemetryParameters.deepLinkPath: path
            ]
        )
    }

    func logAIRequest(type: String, model: String, input: String? = nil) {
        logKpknEvent(
            eventName: TelemetryEvents.aiRequest,
            params: [
                TelemetryParameters.aiType: type,
                TelemetryParameters.aiModel: model,
                TelemetryParameters.aiInput: input.map { anonymizeText($0) } as Any
            ]
        )
    }

    func logAISuccess(type: String, model: String, output: String? = nil) {
        logKpknEvent(
            eventName: TelemetryEvents.aiSuccess,
            params: [
                TelemetryParameters.aiType: type,
                TelemetryParameters.aiModel: model,
                TelemetryParameters.aiOutput: output.map { anonymizeText($0) } as Any,
                TelemetryParameters.success: true
            ]
        )
    }

    func logAIFailed(type: String, model: String, error: String? = nil) {
        logKpknEvent(
            eventName: TelemetryEvents.aiFailed,
            params: [
                TelemetryParameters.aiType: type,
                TelemetryParameters.aiModel: model,
                TelemetryParameters.errorMessage: error as Any,
                TelemetryParameters.success: false
            ]
        )
    }

    class PerformanceTrace {
        private let telemetry: KpknTelemetry
        private let name: String
        private let startTime: TimeInterval

        init(telemetry: KpknTelemetry, name: String) {
            self.telemetry = telemetry
            self.name = name
            self.startTime = ProcessInfo.processInfo.systemUptime
        }

        func stop() {
            let duration = ProcessInfo.processInfo.systemUptime - startTime
            telemetry.logEvent(
                eventName: "performance_trace",
                params: [
                    TelemetryParameters.action: name,
                    TelemetryParameters.duration: duration,
                    TelemetryParameters.success: true
                ]
            )
        }
    }

    func startPerformanceTrace(name: String) -> PerformanceTrace {
        PerformanceTrace(telemetry: telemetry, name: name)
    }

    func logGlobalError(_ error: Error, context: String? = nil) {
        telemetry.logException(error, fatal: false)
        telemetry.logEvent(
            eventName: "global_error",
            params: [
                TelemetryParameters.errorMessage: error.localizedDescription,
                TelemetryParameters.errorStackTrace: String("\(error)".prefix(500)),
                TelemetryParameters.action: context as Any
            ]
        )
    }
}
